typing/typecore.ml

```diff
@@ -4024,18 +4017,22 @@ and type_expect_
           with
             Rpresent (Some ty), Rpresent (Some ty0) ->
               let arg = type_argument env argument_mode sarg ty ty0 in
-              re { exp_desc = Texp_variant(l, Some arg);
+              let alloc_mode = register_allocation expected_mode in
+              re { exp_desc = Texp_variant(l, Some (arg, alloc_mode));
                    exp_loc = loc; exp_extra = [];
                    exp_type = ty_expected0;
-                   exp_mode = expected_mode.mode;
                    exp_attributes = sexp.pexp_attributes;
                    exp_env = env }
           | _ -> raise Exit
           end
       | _ -> raise Exit
       with Exit ->
-        let arg = Option.map (type_exp env argument_mode) sarg in
-        let arg_type = Option.map (fun arg -> arg.exp_type) arg in
+        let arg = match sarg with
+        | None -> None
+        | Some sarg -> Some (type_exp env argument_mode sarg,
```

Comment:
(Style note: more usual to put a newline after `->` if the case doesn't fit on a single line.)